### PR TITLE
feat: dual-venvs for faster/cheaper environment creation

### PR DESCRIPTION
**Pull request by isidentical**
TBD

### CODE REVIEW

**Comment by chamini2** [comment]
fal has 

```
pandas = "^1.3.4"
```

What happens if a user setups an environment

```yaml
environments:
  - name: funny
    requirements:
      - pandas==0.25.0
```

From the explanation I understood that it would find pandas 0.25.0 always when running because it will give a precedence to the `funny` environment PYTHONPATH.

In case this is like this, can we avoid it by checking (but not installing) package dependency resolution of the main environment (fal,dbt,adapters) and the secondary together? And would it be worth it?

File: src/fal/packages/environments/virtual_env.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            for search_path in search_paths
+        )
+
+        python_executable = primary_path / "bin" / "python"
+        logger.debug("Starting the process...")
+        with subprocess.Popen(
+            [
+                python_executable,
+                isolated_runner.__file__,
+                bridge.encode_service_address(service.address),
```
[the remark this answers is not in the document]
Comment (isidentical) [reply src/fal/packages/environments/virtual_env.py:106]: Done.

File: src/fal/packages/environments/virtual_env.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        # take precedence.
+        search_paths = [secondary_path, primary_path]
+        python_path = os.pathsep.join(
+            # sysconfig takes the virtual environment path and returns
+            # the directory where all the site packages are located.
+            sysconfig.get_path("purelib", vars={"base": search_path})
+            for search_path in search_paths
+        )
+
+        python_executable = primary_path / "bin" / "python"
```
[the remark this answers is not in the document]
Comment (isidentical) [reply src/fal/packages/environments/virtual_env.py:100]: Yes for posix (linux/macos) and no for windows. We don't test it on windows yet, but I'm creating an issue just to keep track of it. https://cs.github.com/adriangb/pdm/blob/8581b5c1c1417d910f9b88008fba678ccd982dde/install-pdm.py#L281. https://linear.app/features-and-labels/issue/FEA-379/decide-on-windows-support-for-environment-management

**Comment by isidentical** [comment]
@chamini2 that's a very nice catch! I think we can do it by simply reusing the PYTHONPATH trick but with `pip` this time. After we construct both environments, we can call `pip check` which ideally works in a single environment where you might have conflicting dependencies but now we can convince it to look for both environments and tell us if there are any problems.

```
 $ PYTHONPATH=/secondary_env:/primary_env /primary_env/bin/pip check
fal 0.5.3a0 has requirement pandas<2.0.0,>=1.3.4, but you have pandas 1.3.3.
```

**Comment by chamini2** [comment]
Sounds great! @isidentical

File: src/fal/packages/environments/base.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        try:
+            status, exception = connection.recv()
+        except EOFError:
+            raise RuntimeError("The child process has unexpectedly exited.")
+
+        if exception is not None:
+            logger.error("An exception has occurred: {}", exception)
+            raise exception
+
+        logger.debug("Isolated process has exitted with status: {}", status)
```
Comment (chamini2) [inline src/fal/packages/environments/base.py:86]:
```suggestion
        logger.debug("Isolated process has exited with status: {}", status)
```

**Review by chamini2 — APPROVED** [review]
!!!

**Pull request merged**